### PR TITLE
Rawprocess in servermode

### PR DESCRIPTION
**Pull request by nkx111**
Large nkx111  [](https://gitlab.cern.ch/rest-for-physics/framework/-/commits/rawprocess-servermode) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=rest-for-physics&utm_campaign=badge_info)

I have implemented a new feature to run raw processes in server mode. We are now able to add input files lively during process through inter-process messenger. restManager will hang up at the end of the last input file waiting for the next input. It will quit only when receiving quit message. This feature is very helpful to have REST cooperation with daq system. For example, as soon as we start data taking, the analysis chain starts, and the figures are drawn in time.

For example, in bash window 1, we launch restManager using example rml: 
```
restManager --c realtimeCoboAnalysis.rml --d 7786 --batch --i "/data2/2MM/M1/graw/CoBo_AsAd*_2020-07-13T13:43:13*_0000.graw" --o abc.root
```
Here `--d 7786` means we define the run number as 7786. REST will run the process and hang up after the file is finished. Then in bash window 2, we run: 
```
restSendMessage 116027 "7786->/data2/2MM/M1/graw/CoBo_AsAd2_2020-07-13T13:43:13.698_0001.graw"
```
Here 116027 is a token for TRestMessenger connection. This command means sending a message telling to add a new file to run 7786. We can see in window 1 the process runs again.
Finally in window 2 we input:
```
restSendMessage 116027 "7786->finish"
```

We can see the process quit normally after all the files has been processed.

image

### CODE REVIEW

**Comment by jgalan** [comment]
Hi @nkx, just a quick question we had with @DavidDiezIb . Does the real time drawing work with unclosed files, or it needs the file to be close in order to update the contents?

**Comment by nkx111** [comment]
> Does the real time drawing work with unclosed files, or it needs the file to be closed in order to update the drawing contents?

If we put TRestRealTimeDrawingProcess in the process chain, then it doesn't need to work with files. It draws png files from process's AnalysisTree. This will need raw processes to be in server mode, as this PR implementes. 

On the other hand, if we setup TRestRealTimeDrawingProcess to be the server, then we don't need to use raw processes in server mode. It needs the file to be closed in order to update the drawing contents. This architecture was implemented, as I reported in the last year's meeting.

**Comment by jgalan** [comment]
There is a compilation problem at the pipeline

**Comment by nkx111** [comment]
it should be gone after this PR

https://github.com/rest-for-physics/connectorslib/pull/4

**Comment by jgalan** [comment]
I see, I imagine if the PR branch at connectors lib would be named `rawprocess-servermode` this pipeline would go green too.

**Comment by nkx111** [comment]
Now it is failed due to TRestRawSignalGeneralFitProcess. Is `ExtractTF1FromString` implemented in another branch in framework?

**Comment by DavidDiezIb** [comment]
This function was renamed to `CreateTF1FromString` here e4b3f751035ee58a9172f4ef464a428c2ad9efac in master branch.

**Comment by jgalan** [comment]
Probably you need to merge `master` on your development branches to update the fix.

**Comment by jgalan** [comment]
However, @DavidDiezIb there is a particular problem with `generalFit` validation pipeline failing. Perhaps it just needs to relax the criteria for pipeline passing?

**Comment by jgalan** [comment]
Pipeline is till getting stuck. This time is at the TREX-DM level. It seems the detector parameters are not properly added to the `TRestDetector`

**Comment by nkx111** [comment]
Now the pipeline is passed!

**Review by jgalan — APPROVED** [review]
This is very useful, probably it would be good to write a page with few indications on how to use this functionality at [rest-for-physics.github.io ](rest-for-physics.github.io ). Just the description at this PR would already be very useful.

File: source/framework/core/src/TRestEventProcess.cxx [hunk 1]
```diff
@@ -68,6 +68,7 @@ TRestEventProcess::TRestEventProcess() {
     fObservablesDefined.clear();
     fObservablesUpdated.clear();
     fSingleThreadOnly = false;
+    fCanvasSize = TVector2(800, 600);
```
[the remark this answers is not in the document]
Comment (nkx111) [reply source/framework/core/src/TRestEventProcess.cxx:71]: This is the default value of `fCanvasSize`. We use automatic parameter loading in some processes. Then we should also set initial value of this datamamber.

[the remark this answers is not in the document]
Comment (nkx111) [reply source/framework/core/src/TRestEventProcess.cxx:71]: for `TRestManagerGUI`, it probably be another package which may use some GUI libraries like QT.

[the remark this answers is not in the document]
Comment (jgalan) [reply source/framework/core/src/TRestEventProcess.cxx:71]: I see it more inside the framework, not as a package, it is more like `restManager`.

**Review by jgalan — APPROVED** [review]

**Review by DavidDiezIb — APPROVED** [review]

**Pull request merged**